Source/Core/InputCommon/ControllerInterface/Pipes/Pipes.cpp

```diff
@@ -0,0 +1,163 @@
+// Copyright 2015 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <iostream>
+#include <string>
+#include <vector>
+#include <map>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <cstdlib>
+
+#include "InputCommon/ControllerInterface/Pipes/Pipes.h"
+#include "Common/FileUtil.h"
+#include "Common/StringUtil.h"
+
+namespace ciface
+{
+namespace Pipes
+{
+static std::vector<std::string> s_button_tokens {
+	"A",
+	"B",
+	"X",
+	"Y",
+	"Z",
+	"START",
+	"L",
+	"R",
+	"D_UP",
+	"D_DOWN",
+	"D_LEFT",
+	"D_RIGHT"
+};
+
+static std::vector<std::string> s_shoulder_tokens{
+	"L",
+	"R"
+};
+
+static std::vector<std::string> s_axis_tokens{
+	"MAIN",
+	"C"
+};
+
+void Init(std::vector<Core::Device*> &devices)
+{
+	// Search the Pipes directory for files that we can open in read-only,
+	// non-blocking mode. The device name is the virtual name of the file.
+	File::FSTEntry fst;
+	int found = 0;
+	std::string dir_path = File::GetUserPath(D_PIPES_IDX);
+	if (!File::Exists(dir_path))
+		return;
+	fst = File::ScanDirectoryTree(dir_path, false);
+	if (!fst.isDirectory)
+		return;
+	for (unsigned int i = 0; i < fst.size; ++i) {
+		File::FSTEntry child = fst.children[i];
+		if (child.isDirectory)
+			continue;
+		int fd = open(child.physicalName.c_str(), O_RDONLY | O_NONBLOCK);
+		if (fd < 0)
+			continue;
+		devices.push_back(new PipeDevice(fd, child.virtualName, found++));
+	}
+}
+
+PipeDevice::PipeDevice(int fd, const std::string &name, int id)
+	: m_fd(fd), m_name(name), m_id(id), m_buf("")
+{
+	for (const auto& tok : s_button_tokens) {
+		PipeInput* btn = new PipeInput("Button " + tok);
+		AddInput(btn);
+		m_buttons[tok] = btn;
+	}
+	for (const auto& tok : s_shoulder_tokens) {
+		AddAxis(tok, 0.0);
+	}
+	for (const auto& tok : s_axis_tokens) {
+		AddAxis(tok + " X", 0.5);
+		AddAxis(tok + " Y", 0.5);
+	}
+}
+
+PipeDevice::~PipeDevice()
+{
+	close(m_fd);
+}
+
+void PipeDevice::UpdateInput()
+{
+	// Read any pending characters off the pipe. If we hit a newline,
+	// then dequeue a command off the front of m_buf and parse it.
+	char buf[32];
+	ssize_t bytes_read = read(m_fd, buf, sizeof buf);
+	while (bytes_read > 0) {
+		m_buf.append(buf, bytes_read);
+		bytes_read = read(m_fd, buf, sizeof buf);
+	}
+	std::size_t newline = m_buf.find_first_of("\n");
+	while (newline != std::string::npos) {
+		std::string command = m_buf.substr(0, newline);
+		ParseCommand(command);
+		m_buf.erase(0, newline + 1);
```

Comment:
I doubt performance matters, but this line end in N² complexity for lots of input in the queue. It would likely be better to parse the input on every read but to try to read everything first.